hy/compiler.py

```diff
@@ -382,6 +387,13 @@ def is_unpack(kind, x):
             and x[0] == "unpack_" + kind)
 
 
+def is_builtin(kind, x):
+    return (isinstance(x, HyExpression)
+            and len(x) > 0
+            and isinstance(x[0], HySymbol)
+            and x[0] in (kind, "__builtin__" + kind))
```

Comment:
Shouldn't this be `x[0] == "__builtins__." + kind`? You don't want to allow `kind` alone because then you'll trigger on `slice` alone, which might be rebound. Also, you need the period.

Reply:
Well `__builtins__.slice` doesn't work because __builtins__ is a dictionary, not an object. So we'd have to either change that too or write `(get __builtins__ "slice")`. And I'm not particularly eager to use throw that large expression down every time I want a slice to be slightly faster (and to be honest, same for `__builtins__.slice`).

If we don't feel comfortable with rewriting `slice`, I can just drop that entire part of the PR.

Reply:
> Well `__builtins__.slice` doesn't work because `__builtins__` is a dictionary, not an object.

Are you sure?

    $ python2
    Python 2.7.14 (default, Sep 23 2017, 22:06:14) 
    [GCC 7.2.0] on linux2
    Type "help", "copyright", "credits" or "license" for more information.
    >>> __builtins__.slice
    <type 'slice'>

    $ python3
    Python 3.6.3 (default, Oct  3 2017, 21:45:48) 
    [GCC 7.2.0] on linux
    Type "help", "copyright", "credits" or "license" for more information.
    >>> __builtins__.slice
    <class 'slice'>

Reply:
hy complains though:

```
=> (cut [1 2 3] 1 2 3)
Traceback (most recent call last):
  File "/home/simon/src/hy/hy/importer.py", line 201, in hy_eval
    return eval(ast_compile(expr, "<eval>", "eval"), namespace)
  File "<eval>", line 1, in <module>
AttributeError: 'dict' object has no attribute 'slice'
```

Reply:
Huh, that's a bizarre bug. We don't set `__builtins__` anywhere in our code and yet Hy still sees a different object than Python does.

Reply:
[Oh shoot, that's a feature.](https://docs.python.org/3/reference/executionmodel.html) Since we hope to be setting `__main__` appropriately someday, we can't safely use `get` nor attribute lookup. I'm afraid I'm not sure what design would make sense, then.

Reply:
I'm honestly quite comfortable letting slice mean slice here and just documenting the behaviour. We already prevent `slice` from being assigned to. Overriding `__builtins__.slice` in Python also won't change the behaviour of the `:` operator.

I could also add a check to make sure no `slice` macro has been defined and/or add `slice*` to force the construction of a slice object and that probable covers 99% of weird use cases.

Reply:
> We already prevent `slice` from being assigned to.

Are you sure?

     hy 0.13.0+211.g5c34296 using CPython(default) 3.6.3 on Linux
     => (setv slice "foo")
     => slice
     'foo'

Anyway, if we want to make `get` recognize a magic symbol, why not use `:` rather than `slice`? So you'd write e.g. `(get foo (: 3 4))` to mean `foo[3:4]`. This is concise, and literal keywords are already magical as function arguments.

Reply:
I do like this better. My concern was that `:` is `HyKeyword(":")`, but there's less room for things to blow up.

> Are you sure?

Yes your right.

I think I was playing `@builds("slice")` at the time I tested it and forgot that I had put that in there.